### PR TITLE
feat(query-params): proper handling in case of invalid credentials

### PR DESCRIPTION
**Pull request by axe312ger**
* Set cookie only if credentials are valid
* Catch invalid credentials and show settings page in that case
* Validate credentials also on GET /settings
* Construct query string with credentials for modal only when credentials are valid

### CODE REVIEW

**Comment by Khaledgarbaya** [comment]
@axe312ger should we probably wait a bit until we release the other part

File: routes/index.js [hunk 1]
```diff
@@ -12,6 +12,15 @@ const { getSettings, postSettings } = require('./settings')
 const { getLandingPage } = require('./landingPage')
 const { getImprint } = require('./imprint')
 
+// Display settings in case of invalid credentials
+router.all('*', async (request, response, next) => {
+  if (response.locals.forceSettingsRoute) {
+    await getSettings(request, response, next)
```
Comment (axe312ger) [inline routes/index.js]: Can we do an `next('/settings')` here? 🤔

Comment (Khaledgarbaya) [reply routes/index.js]: nope if you pass anything to next it will be treated as an error

Comment (Khaledgarbaya) [reply routes/index.js]: we can do response.redirect('/settings')

Comment (axe312ger) [reply routes/index.js]: Which won't work because headers have been sent already.

File: helpers.js [hunk 2]
```diff
@@ -29,6 +31,38 @@ module.exports.isCustomCredentials = (settings) => {
...
+    editorial_features: settings.editorialFeatures ? 'enabled' : 'disabled'
+  })
+
+  if (isCustomCredentials(settings)) {
+    settingsQuery = Object.assign(settingsQuery, {
+      space_id: settings.spaceId,
+      delivery_token: settings.deliveryToken,
+      preview_token: settings.previewToken
+    })
+  }
```
Comment (loudmouth) [inline helpers.js]: So a little confused about the above two functions:
- the `cleanupQueryParameters` function seemingly strips a query of everything...why not just initialize an empty query?
- Then, we take a newly created query and populate the settings with what was in already stored in our session. Is my understanding correct? The spec says we should populate the fields on the settings screen with the incorrect credentials (but of course showing the validation errors to the user, which i'll assume is done).

File: routes/settings.js [hunk 3]
```diff
@@ -2,24 +2,24 @@
...
   // Get connected space to display the space name on top of the settings
   let space = false
   try {
     space = await getSpace()
   } catch (error) {
-    // We throw the error here, it will be handled by the error middleware
-    // We keep space false to ensure the "Connected to" box is not shown.
-    throw (error)
+    // We handle errors within the settings page.
+    // No need to throw here.
```
Comment (loudmouth) [inline routes/settings.js]: 👍

File: views/settings.pug [hunk 4]
```diff
@@ -64,7 +64,7 @@ block content
                   br
                   button(type="submit") #{translate('resetCredentialsLabel', currentLocale.code)}
                   br
-                  a(href=`${baseUrl}/?space_id=${settings.spaceId}&preview_token=${settings.previewToken}&delivery_token=${settings.deliveryToken}&api=${currentApi.id}${settings.editorialFeatures ? '&enable_editorial_features' : ''}` class="status-block__sharelink") #{translate('copyLinkLabel', currentLocale.code)}
+                  a(href=`${baseUrl}/${queryStringSettings}` class="status-block__sharelink") #{translate('copyLinkLabel', currentLocale.code)}
```
Comment (loudmouth) [inline views/settings.pug]: 👍

**Comment by loudmouth** [comment]
@axe312ger I had one question, but it looked good overall ;-)

File: helpers.js (hunk 2 above)
Comment (axe312ger) [reply helpers.js]: The function keeps api & locale setting in the query. Thats why an empty object is not a solution.

Plus, our session storage is independent form the query. We will try to apply the credentials from the query, store them in the session when they are valid, but never keep the credentials in the query.

**Pull request merged**